grub-core/loader/emu/linux.c

```diff
@@ -33,6 +32,201 @@ static char *kernel_path;
+      grub_dprintf ("linux", "switch_root: No kernel_path found\n");
+      goto out;
+    }
+
+  /* Check for kernel mismatch  */
+  /* Retrieve the current kernel relase tag */
+  grub_util_exec_redirect (unamer, NULL, "/tmp/version");
+
+  grub_file_t f = grub_file_open ("/tmp/version", GRUB_FILE_TYPE_FS_SEARCH);
+
+  if (f == NULL)
+    {
+      grub_dprintf ("linux", "failed opening file.\n");
+      rc = GRUB_ERR_FILE_NOT_FOUND;
+      goto out;
+    }
+  else
+    {
+      if ((uname_buf = grub_malloc (f->size)) == NULL)
+        {
+          grub_dprintf ("linux", "switch_root: Failed to allocate memory\n");
+          rc = GRUB_ERR_OUT_OF_MEMORY;
+          goto out;
+        }
+
+	  if (grub_file_read (f, uname_buf, f->size) < 0)
+      {
+        grub_dprintf ("linux", "switch_root: failed to read from file\n");
+        rc = GRUB_ERR_FILE_READ_ERROR;
+        goto out;
+      }
+
+    grub_file_close (f);
+
+    if (grub_strstr (uname_buf, kernel_release) == NULL)
+      {
+        grub_dprintf ("linux", "switch_root: kernel mismatch, aborting ...\n");
+        rc = GRUB_ERR_NO_KERNEL;
+        goto out;
+      }
+    }
+
+  /* Extract the root partition from boot_cmdline */
+  if (boot_cmdline)
+    {
+      int i = 0;
+      char* tmp = grub_strdup (boot_cmdline);
+
+      if (tmp == NULL) {
+        rc = GRUB_ERR_OUT_OF_MEMORY;
+        goto out;
+      }
+
+      if ((root_uuid = grub_strstr (tmp, root_prefix)) == NULL)
+        {
+          rc = GRUB_ERR_BAD_ARGUMENT;
+          grub_dprintf ("linux", "switch_root: Can't find rootfs\n");
+          grub_free (tmp);
+          goto out;
+        }
+
+      root_uuid += grub_strlen (root_prefix);
+
+      while ((root_uuid[i] != ' ') && (root_uuid[i] != '\0'))
+        i ++;
+      root_uuid[i] = '\0';
+
+      /* Allocate a new buffer holding root_uuid*/
+      root_uuid = grub_xasprintf ("%s", root_uuid);
+
+      if (root_uuid == NULL)
+        {
+          grub_dprintf ("linux", "switch_root: Failed to allocated memory\n");
+          grub_free (tmp);
+          rc = GRUB_ERR_OUT_OF_MEMORY;
+          goto out;
+        }
+
+      /* Check for subvol parameter */
+      grub_strcpy (tmp, boot_cmdline);
+
+      if ((subvol = grub_strstr(tmp, subvol_param)) != NULL)
+        {
+          i = 0;
+
+          while ((subvol[i] != ' ') && (subvol[i] != '\0'))
+            i ++;
+          subvol[i] = '\0';
+
+          /* Allocate a new buffer holding subvol */
+          subvol = grub_xasprintf("%s", subvol);
+
+          if (subvol == NULL)
+            {
+              grub_dprintf ("linux", "switch_root: Failed to allocated memory\n");
+              grub_free (tmp);
+              rc = GRUB_ERR_OUT_OF_MEMORY;
+              goto out;
+            }
+
+          options_cmd = grub_xasprintf("%s", "-o");
+          options = grub_xasprintf("%s", subvol);
+        }
+
+      grub_free (tmp);
+    }
+  else
+    {
+      rc = GRUB_ERR_BAD_ARGUMENT;
+      goto out;
+    }
```

Comment:
So here and in a couple other places, there's now code of the form:

```C
if (foo)
  {
    /* lots of stuff */
  }
else
  {
     rc = failure;
     goto out;
   }

/* more code here */

out:
  ...
```

This can be reformed as:

```C
if (!foo)
  {
    rc = failure;
    goto out;
  }

/* lots of stuff */
/* more code here */
out:
  ...
```

if that makes sense.